File: blur.py
"""
@Chatbot AI
@Mopheshi
"""

import cv2
import numpy as np

# Load the img
img = cv2.imread("Images/l'agertha.jpg")

# Convert to grayscale
gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

# Create a kernel The size of the kernel affects the amount of blurring. A larger kernel size will result in stronger
# blurring. However, increasing the kernel size also increases the execution time of the algorithm.
kernel_size = 5
kernel = np.ones((kernel_size, kernel_size), np.float32)
kernel = kernel / (kernel_size * kernel_size)

# Convolution
result = cv2.filter2D(gray, -1, kernel)

# Convert grayscale to RGB
# blurredImage = cv2.cvtColor(img, cv2.COLOR_GRAY2RGB)

# Save the blurred img
cv2.imwrite("Images/blurred_image.jpg", result)

# Show blurred img
cv2.imshow("Original Image", img)
cv2.imshow("Blurred Image", result)
# waitKey = 0 means the system will continue to show img until user presses any key
cv2.waitKey(0)

File: Blur.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package exercises;

import java.awt.image.BufferedImage;
import java.io.File;
import javax.imageio.ImageIO;

/**
 *
 * @author Chatbot AI
 * @author MOPHE
 */
public class Blur {

    public static void main(String[] args) throws Exception {
        final String importPath = "image.jpg";
        final String exportPath = "blurredImage.jpg";

        // Load the image
        BufferedImage img = ImageIO.read(new File(importPath));
        int width = img.getWidth();
        int height = img.getHeight();

        // Create a kernel
        // Note: The performance of the algorithm may decrease for larger images or kernel sizes,
        // so some optimization may be necessary to improve execution time.
        int kernelSize = 5;
        double[][] kernel = new double[kernelSize][kernelSize];
        double kernelSum = 0.0;
        for (int i = 0; i < kernelSize; i++) {
            for (int j = 0; j < kernelSize; j++) {
                kernel[i][j] = 1.0 / (kernelSize * kernelSize);
                kernelSum += kernel[i][j];
            }
        }

        // Convert the image to a two-dimensional array of integers
        int[][] pixels = new int[width][height];
        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                pixels[x][y] = img.getRGB(x, y) & 0xff;
            }
        }

        // Apply the convolution operation using the kernel matrix
        int[][] newPixels = new int[width][height];
        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                double pixelSum = 0.0;
                for (int i = 0; i < kernelSize; i++) {
                    for (int j = 0; j < kernelSize; j++) {
                        int pixelVal = getPixelValue(pixels, width, height, x + i, y + j);
                        pixelSum += kernel[i][j] * pixelVal;
                    }
                }
                newPixels[x][y] = (int) (pixelSum / kernelSum);
            }
        }

        // Replace the original pixels with the new blurred pixels
        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                int p = (newPixels[x][y] << 16) | (newPixels[x][y] << 8) | newPixels[x][y];
                img.setRGB(x, y, p);
            }
        }

        // Save the blurred image
        ImageIO.write(img, "jpg", new File(exportPath));
    }

    // Get the pixel value of the image, handling out of bounds conditions
    private static int getPixelValue(int[][] pixels, int width, int height, int x, int y) {
        if (x < 0 || x >= width || y < 0 || y >= height) {
            return 0;
        }
        return pixels[x][y];
    }
}
